libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h

```diff
@@ -0,0 +1,525 @@
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment,
+    typename = void>
+struct __invoke_if_applicable {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
+        return _CUDA_VSTD::forward<_Fn>(__f)(__alignment<_Alignment>());
+    }
+};
+
+template<_CUDA_VSTD::size_t _GuaranteedAlignment,
+    _CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment>
+struct __invoke_if_applicable<_GuaranteedAlignment,
+    _MaxInterestingAlignment,
+    _MinInterestingAlignment,
+    _Alignment,
+    _CUDA_VSTD::__enable_if_t<
+        _Alignment >= _MaxInterestingAlignment
+            || _Alignment < _MinInterestingAlignment
+            || _Alignment < _GuaranteedAlignment
+>> {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
+        _LIBCUDACXX_UNREACHABLE();
+    }
+};
+
+template<_CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _GuaranteedAlignment,
+    typename _Fn>
+_LIBCUDACXX_INLINE_VISIBILITY async_contract_fulfillment __dispatch_alignment_bit(_Fn && __f, _CUDA_VSTD::size_t __alignment_fsb) {
+    _CUDA_VSTD::size_t __alignment_v = 1ull << (__alignment_fsb - 1);
+
+    if (__alignment_v >= _MaxInterestingAlignment) [[likely]] {
+        return _CUDA_VSTD::forward<_Fn>(__f)(__alignment<_MaxInterestingAlignment>());
+    }
+
+    if (__alignment_v < _MinInterestingAlignment) [[unlikely]] {
+        return __invoke_if_applicable<_GuaranteedAlignment, _MaxInterestingAlignment, 0, 1>::__invoke(_CUDA_VSTD::forward<_Fn>(__f));
+    }
+
+    switch (__alignment_fsb) {
+#define _ADD_CASE(val)                                                                                      \
+    case val:                                                                                               \
+        return __invoke_if_applicable<                                                                      \
+            _GuaranteedAlignment,                                                                           \
+            _MaxInterestingAlignment,                                                                       \
+            _MinInterestingAlignment,                                                                       \
+            1ull << (val - 1)>::__invoke(_CUDA_VSTD::forward<_Fn>(__f))
+
+        _ADD_CASE(12);
+        _ADD_CASE(11);
+        _ADD_CASE(10);
+        _ADD_CASE(9);
+        _ADD_CASE(8);
+        _ADD_CASE(7);
+        _ADD_CASE(6);
+        _ADD_CASE(5);
+        _ADD_CASE(4);
+        _ADD_CASE(3);
+        _ADD_CASE(2);
+        _ADD_CASE(1);
+
+#undef _ADD_CASE
+    }
+
+    _LIBCUDACXX_UNREACHABLE();
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY _CUDA_VSTD::size_t __get_size(_CUDA_VSTD::size_t __size) {
```

Comment:
Why are those functions not `constexpr`?

Reply:
I have a local patch that turns them into not-even-actual-functions, so this isn't relevant anymore.